plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/autoedit/JSDocEditStrategy.java

```diff
@@ -0,0 +1,142 @@
+			IRegion stopTerminal) throws BadLocationException {
+		CommandInfo newC = new CommandInfo();
+		List<String> returnTypeAndParameterNames = getReturnTypeAndParameterNames(document, startTerminal);
+		String paramString = "";
+		String returnString = "";
+		if ((returnTypeAndParameterNames.size() > 0) && returnTypeAndParameterNames.get(0).equals("return")) {
+			returnString = INDENTATION_STR + RETURN_STR + command.text;
+		}
+		if (returnTypeAndParameterNames.size() > 1) {
+			for (int i = 1; i < returnTypeAndParameterNames.size(); i += 1) {
+				paramString += command.text + INDENTATION_STR + PARAM_STR + returnTypeAndParameterNames.get(i);
+			}
+		}
+		newC.isChange = true;
+		newC.offset = command.offset;
+		newC.text += command.text + INDENTATION_STR;
+		newC.cursorOffset = command.offset + newC.text.length();
+		if (stopTerminal == null && atEndOfLineInput(document, command.offset)) {
+			newC.text += command.text + getRightTerminal();
+		}
+		if (stopTerminal != null && stopTerminal.getOffset() >= command.offset
+				&& util.isSameLine(document, stopTerminal.getOffset(), command.offset)) {
+			String string = document.get(command.offset, stopTerminal.getOffset() - command.offset);
+			if (string.trim().length() > 0)
+				newC.text += string.trim();
+			if (!(returnTypeAndParameterNames.size() == 0)) {
+				newC.text += paramString + command.text + returnString;
+			} else {
+				newC.text += command.text;
+			}
+
+			newC.length += string.length();
+		}
+		return newC;
+	}
+
+	/**
+	 * @returns List<String> otherwise null in case the document is not N4JSDocument or something breaks
+	 */
+	private List<String> getReturnTypeAndParameterNames(IDocument document, IRegion startTerminal) {
+		List<String> retAndfparNames = null;
+		if (document instanceof N4JSDocument) {
+			// index 0: if set to "void", the method has no return value;
+			// fpars start at index 1
+			retAndfparNames = ((N4JSDocument) document).tryReadOnly((xtextResource) -> {
+				Script script = XtextResourceUtils.getScript(xtextResource);
+				ILeafNode node = NodeModelUtils.findLeafNodeAtOffset(NodeModelUtils.findActualNodeFor(script),
+						startTerminal.getOffset());
+				EObject astElement = node.getSemanticElement();
+				List<String> retAndFPars = new ArrayList<>();
+				if (astElement instanceof LiteralOrComputedPropertyName) {
+					if (astElement.eContainer() instanceof N4MethodDeclaration) {
+						N4MethodDeclaration methodDecl = (N4MethodDeclaration) astElement.eContainer();
+						getRetAndFPars(retAndFPars, methodDecl);
+					}
+				} else if (astElement instanceof N4MethodDeclaration) {
+					N4MethodDeclaration methodDecl = (N4MethodDeclaration) astElement;
+					getRetAndFPars(retAndFPars, methodDecl);
+				}
+				return retAndFPars;
+			});
+		}
+		return retAndfparNames;
+	}
+
+	private void getRetAndFPars(List<String> retAndFPars, N4MethodDeclaration methodDecl) {
```

Comment:
Add comment:

```
The first element of the list (index 0) marks whether the method has a result (return type) or not. In the first case, "return" is found at index 0, otherwise "void". The parameter names are added after that, i.e. starting at index 1.
```